File: src/main/java/com/imbananko/tilly/MemeManager.java
package com.imbananko.tilly;

import com.imbananko.tilly.model.MemeEntity;
import com.imbananko.tilly.model.VoteEntity;
import com.imbananko.tilly.repository.MemeRepository;
import com.imbananko.tilly.repository.VoteRepository;
import com.imbananko.tilly.similarity.MemeMatcher;
import com.imbananko.tilly.utility.TelegramPredicates;
import io.vavr.collection.HashMap;
import io.vavr.control.Try;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.io.IOUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.bots.TelegramLongPollingBot;
import org.telegram.telegrambots.meta.api.methods.GetFile;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
import org.telegram.telegrambots.meta.api.methods.send.SendPhoto;
import org.telegram.telegrambots.meta.api.methods.updatingmessages.EditMessageReplyMarkup;
import org.telegram.telegrambots.meta.api.objects.File;
import org.telegram.telegrambots.meta.api.objects.Update;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.InlineKeyboardMarkup;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.InlineKeyboardButton;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;

import javax.annotation.PostConstruct;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.List;
import java.util.Optional;
import java.util.function.BiFunction;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Supplier;

import static com.imbananko.tilly.model.VoteEntity.Value.*;
import static io.vavr.API.*;
import static io.vavr.Patterns.$None;
import static io.vavr.Patterns.$Some;
import static io.vavr.Predicates.allOf;

@Slf4j
@Component
public class MemeManager extends TelegramLongPollingBot {
  private final MemeRepository memeRepository;
  private final VoteRepository voteRepository;

  private final MemeMatcher memeMatcher;

  @Value("${target.chat.id}")
  private long chatId;

  @Value("${bot.token}")
  private String token;

  @Value("${bot.username}")
  private String username;

  @Autowired
  public MemeManager(MemeRepository memeRepository, VoteRepository voteRepository, MemeMatcher memeMatcher) {
    this.memeRepository = memeRepository;
    this.voteRepository = voteRepository;
    this.memeMatcher = memeMatcher;
  }

  @PostConstruct
  public void init() {
    memeRepository
        .load(chatId)
        .forEach(me -> {
          try {
            memeMatcher.addMeme(me.getFileId(), downloadFromFileId(me.getFileId()));
          } catch (TelegramApiException | IOException e) {
            log.error("Failed to load file {}: {}, skipping...", me.getFileId(), e.getMessage());
          }
        });
  }

  @Override
  public String getBotToken() {
    return token;
  }

  @Override
  public String getBotUsername() {
    return username;
  }

  @Override
  public void onUpdateReceived(Update update) {
    Match(update)
      .of(
        Case($(allOf(TelegramPredicates.isP2PChat(), TelegramPredicates.hasPhoto())), it -> run(() -> this.processMeme(it))),
        Case($(TelegramPredicates.hasVote()), it -> run(() -> this.processVote(it))),
        Case($(), () -> null)
      );
  }

  private void processMeme(Update update) {
    final var message = update.getMessage();
    final var fileId = message.getPhoto().get(0).getFileId();
    final var authorUsername = message.getFrom().getUserName();

    final var memeCaption =
      Optional.ofNullable(message.getCaption()).map(it -> it.trim() + "\n\n").orElse("")
        + "Sender: "
        + Optional.ofNullable(authorUsername).orElse("tg://user?id=" + message.getFrom().getId());

    final Supplier<Try> processMemeIfUnique = () -> Try.of(() -> execute(
      new SendPhoto()
        .setChatId(chatId)
        .setPhoto(fileId)
        .setCaption(memeCaption)
        .setReplyMarkup(createMarkup(HashMap.empty(), false))
      )
    )
      .onSuccess(sentMemeMessage -> {
        final var meme = MemeEntity.builder()
          .chatId(sentMemeMessage.getChatId())
          .messageId(sentMemeMessage.getMessageId())
          .senderId(message.getFrom().getId())
          .fileId(message.getPhoto().get(0).getFileId())
          .build();
        log.info("Sent meme=" + meme);
        memeRepository.save(meme);
      })
        .onFailure(throwable -> log.error("Failed to send meme from message=" + message + ". Exception=" + throwable.getCause()));

    final Function<String, Try> processMemeIfExists = (existingMemeId) -> Try.of(() -> execute(
        new SendPhoto()
            .setChatId(chatId)
            .setPhoto(fileId)
            .setCaption(String.format("@%s попытался отправить этот мем, несмотря на то, что его уже скидывали выше. Позор...", authorUsername))
            .setReplyToMessageId(memeRepository.messageIdByFileId(existingMemeId, chatId))
    )).onFailure(throwable -> log.error("Failed to reply with existing meme from message=" + message + ". Exception=" + throwable.getCause()));

    Try.of(() -> downloadFromFileId(fileId))
        .flatMap(memeFile -> memeMatcher.checkMemeExists(fileId, memeFile))
        .flatMap(memeId -> Match(memeId).of(
            Case($Some($()), processMemeIfExists),
            Case($None(), processMemeIfUnique)
        ))
        .onFailure(new Consumer<Throwable>() {
          @Override
          public void accept(Throwable throwable) {
            log.error("Failed to check if meme is unique, sending anyway. Exception={}", throwable.getMessage());
            processMemeIfUnique.get();
          }
        });
  }

  private void processVote(Update update) {
    final var message = update.getCallbackQuery().getMessage();
    final var targetChatId = message.getChatId();
    final var messageId = message.getMessageId();
    final var vote = VoteEntity.Value.valueOf(update.getCallbackQuery().getData().split(" ")[0]);
    final var voteSender = update.getCallbackQuery().getFrom();
    final var memeSenderFromCaption = message.getCaption().split("Sender: ")[1];

    final var wasExplained = message
      .getReplyMarkup()
      .getKeyboard().get(0).get(1)
      .getCallbackData()
      .contains("EXPLAINED");

    var voteEntity =
      VoteEntity.builder()
        .chatId(targetChatId)
        .messageId(messageId)
        .voterId(voteSender.getId())
        .value(vote)
        .build();

    if (voteSender.getUserName().equals(memeSenderFromCaption) || memeRepository.getMemeSender(targetChatId, messageId).equals(voteSender.getId())) {
      return;
    }

    if (voteRepository.exists(voteEntity)) {
      voteRepository.delete(voteEntity);
    } else {
      voteRepository.insertOrUpdate(voteEntity);
    }

    final var statistics = voteRepository.getStats(targetChatId, messageId);
    final var shouldMarkExplained = vote.equals(EXPLAIN) && !wasExplained && statistics.getOrElse(EXPLAIN, 0L) == 3L;

    Try.of(() -> execute(
      new EditMessageReplyMarkup()
        .setMessageId(messageId)
        .setChatId(targetChatId)
        .setInlineMessageId(update.getCallbackQuery().getInlineMessageId())
        .setReplyMarkup(createMarkup(statistics, wasExplained || shouldMarkExplained))
      )
    )
      .onSuccess(ignore -> log.info("Processed vote=" + voteEntity))
      .onFailure(throwable -> log.error("Failed to process vote=" + voteEntity + ". Exception=" + throwable.getMessage()));

    if (shouldMarkExplained) {

      final var replyText =
        update.getCallbackQuery().getMessage().getCaption().replaceFirst("Sender: ", "@")
          + ", поясни за мем";

      Try.of(() -> execute(
        new SendMessage()
          .setChatId(targetChatId)
          .setReplyToMessageId(update.getCallbackQuery().getMessage().getMessageId())
          .setText(replyText)
        )
      )
        .onSuccess(ignore -> log.info("Successful reply for explaining"))
        .onFailure(throwable -> log.error("Failed to reply for explaining. Exception=" + throwable.getMessage()));
    }
  }

  private static InlineKeyboardMarkup createMarkup(HashMap<VoteEntity.Value, Long> stats, boolean markExplained) {
    BiFunction<VoteEntity.Value, Long, InlineKeyboardButton> createVoteInlineKeyboardButton = (voteValue, voteCount) -> {
      final var callbackData = voteValue.equals(EXPLAIN) && markExplained
              ? voteValue.name() + " EXPLAINED"
              : voteValue.name();

      return new InlineKeyboardButton()
              .setText(voteCount == 0L ? voteValue.getEmoji() : voteValue.getEmoji() + " " + voteCount)
              .setCallbackData(callbackData);
    };

    return new InlineKeyboardMarkup().setKeyboard(
      List.of(
        List.of(
          createVoteInlineKeyboardButton.apply(UP, stats.getOrElse(UP, 0L)),
          createVoteInlineKeyboardButton.apply(EXPLAIN, stats.getOrElse(EXPLAIN, 0L)),
          createVoteInlineKeyboardButton.apply(DOWN, stats.getOrElse(DOWN, 0L))
        )
      )
    );
  }

  private java.io.File downloadFromFileId(String fileId) throws TelegramApiException, IOException {
    GetFile getFile = new GetFile();
    getFile.setFileId(fileId);

    File file = execute(getFile);
    URL fileUrl = new URL(file.getFileUrl(getBotToken()));
    HttpURLConnection httpConn = (HttpURLConnection) fileUrl.openConnection();
    InputStream inputStream = httpConn.getInputStream();

    final java.io.File tempFile = java.io.File.createTempFile("telegram-photo-", "");
    tempFile.deleteOnExit();
    try (FileOutputStream out = new FileOutputStream(tempFile)) {
      IOUtils.copy(inputStream, out);
    }

    inputStream.close();
    httpConn.disconnect();

    return tempFile;
  }
}

File: src/main/kotlin/com/imbananko/tilly/MemeManager.kt
package com.imbananko.tilly

import com.imbananko.tilly.model.MemeEntity
import com.imbananko.tilly.model.MemeStatsEntry
import com.imbananko.tilly.model.VoteEntity
import com.imbananko.tilly.model.VoteValue
import com.imbananko.tilly.model.VoteValue.DOWN
import com.imbananko.tilly.model.VoteValue.UP
import com.imbananko.tilly.repository.MemeRepository
import com.imbananko.tilly.repository.VoteRepository
import com.imbananko.tilly.similarity.MemeMatcher
import com.imbananko.tilly.utility.*
import org.apache.commons.io.IOUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.scheduling.annotation.EnableScheduling
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import org.telegram.telegrambots.bots.TelegramLongPollingBot
import org.telegram.telegrambots.meta.api.methods.ForwardMessage
import org.telegram.telegrambots.meta.api.methods.GetFile
import org.telegram.telegrambots.meta.api.methods.ParseMode
import org.telegram.telegrambots.meta.api.methods.groupadministration.GetChatMember
import org.telegram.telegrambots.meta.api.methods.pinnedmessages.PinChatMessage
import org.telegram.telegrambots.meta.api.methods.send.SendMediaGroup
import org.telegram.telegrambots.meta.api.methods.send.SendMessage
import org.telegram.telegrambots.meta.api.methods.send.SendPhoto
import org.telegram.telegrambots.meta.api.methods.updatingmessages.DeleteMessage
import org.telegram.telegrambots.meta.api.methods.updatingmessages.EditMessageCaption
import org.telegram.telegrambots.meta.api.methods.updatingmessages.EditMessageReplyMarkup
import org.telegram.telegrambots.meta.api.objects.Update
import org.telegram.telegrambots.meta.api.objects.media.InputMediaPhoto
import org.telegram.telegrambots.meta.api.objects.replykeyboard.InlineKeyboardMarkup
import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.InlineKeyboardButton
import java.io.File.createTempFile
import java.io.FileOutputStream
import java.lang.System.currentTimeMillis
import java.lang.Thread.currentThread
import java.net.URL
import javax.annotation.PostConstruct

@EnableScheduling
@Component
class MemeManager(private val memeRepository: MemeRepository, private val voteRepository: VoteRepository, private val memeMatcher: MemeMatcher) : TelegramLongPollingBot() {

  private val log = LoggerFactory.getLogger(javaClass)

  @Value("\${target.chat.id}")
  private val chatId: Long = 0

  @Value("\${target.channel.id}")
  private val channelId: Long = 0

  @Value("\${bot.token}")
  private lateinit var token: String

  @Value("\${bot.username}")
  private lateinit var username: String

  override fun getBotToken(): String? = token

  override fun getBotUsername(): String? = username

  @PostConstruct
  private fun init() {
    memeRepository
        .findAllChatMemes(chatId)
        .distinctBy { it.fileId }
        .parallelStream()
        .map { meme -> Pair(meme.fileId, downloadFromFileId(meme.fileId)) }
        .forEach { (fileId, file) ->
          runCatching {
            memeMatcher.addMeme(fileId, file)
          }.onFailure { throwable: Throwable ->
            log.error("Failed to load file {}. Exception=", fileId, throwable)
          }
        }
  }

  @Scheduled(cron = "0 0 19 * * WED")
  private fun sendMemeOfTheWeek() {
    runCatching {
      val meme: MemeEntity? = memeRepository.findMemeOfTheWeek(channelId)

      meme ?: return

      val winner = execute(
          GetChatMember()
              .setChatId(channelId)
              .setUserId(meme.senderId))
          .user
      val congratulationText = "Поздравляем ${winner.mention()} с мемом недели!"
      val memeOfTheWeekMessage = runCatching {
        execute(
            SendMessage()
                .setChatId(channelId)
                .setParseMode(ParseMode.MARKDOWN)
                .setReplyToMessageId(meme.channelMessageId)
                .setText(congratulationText)
        )
      }.getOrElse { error ->
        log.error("Can't reply to meme of the week because of", error)
        log.info("Send meme of the week as new message")

        val statistics = voteRepository.getVotes(meme)
            .groupingBy { vote -> vote.voteValue }
            .eachCount()
        val fallbackMemeOfTheWeekMessage = execute(
            SendPhoto()
                .setChatId(channelId)
                .setPhoto(meme.fileId)
                .setParseMode(ParseMode.MARKDOWN)
                .setCaption(congratulationText)
                .setReplyMarkup(createMarkup(statistics))
        )

        memeRepository.update(meme, meme.copy(channelId = meme.channelId, channelMessageId = fallbackMemeOfTheWeekMessage.messageId))

        fallbackMemeOfTheWeekMessage
      }
      execute(PinChatMessage(memeOfTheWeekMessage.chatId, memeOfTheWeekMessage.messageId))
    }
        .onSuccess { log.info("Successful send meme of the week") }
        .onFailure { throwable -> log.error("Can't send meme of the week because of", throwable) }
  }

  @Scheduled(cron = "0 0 20 31 12 *")
  private fun sendMemesOfTheYear() {
    fun formatMemeTheYearCaption(meme: MemeEntity): String {
      val userMention = execute(GetChatMember()
          .setChatId(meme.chatId)
          .setUserId(meme.senderId))
          .user.mention()
      val votes = voteRepository.getVotes(meme)
          .groupingBy { vote -> vote.voteValue }
          .eachCount()
          .map { entry -> entry.key.emoji + " " + entry.value }
          .joinToString(prefix = "(", postfix = ")", separator = ", ")
      return "$userMention $votes"
    }

    runCatching {
      execute(SendMessage(channelId, "Топ мемсы прошедшего года:"))
      execute(
          SendMediaGroup(channelId, memeRepository.findMemesOfTheYear(channelId).map { meme ->
            InputMediaPhoto(meme.fileId, formatMemeTheYearCaption(meme))
                .setParseMode(ParseMode.MARKDOWN)
          })
      )
    }
        .onSuccess { log.info("Successful send memes of the year") }
        .onFailure { throwable -> log.error("Can't send memes of the year because of", throwable) }
  }

  override fun onUpdateReceived(update: Update) {
    if (update.hasMeme()) processMeme(update)
    if (update.hasChannelVote()) processChannelVote(update)
    if (update.hasChatVote()) processChatVote(update)
    if (update.hasStatsCommand()) sendStats(update)
  }

  private fun updateCaptionInSenderChat(meme: MemeEntity, caption: String) =
      runCatching {
        execute(EditMessageCaption()
            .setChatId(meme.senderId.toString())
            .setMessageId(meme.privateMessageId)
            .setCaption(caption)
        )
      }.onFailure {
        log.error("Failed update caption in private chat=$chatId", it)
      }

  private fun processChatVote(update: Update) {
    val messageId = update.callbackQuery.message.messageId

    val meme = memeRepository.findMemeByChat(chatId, messageId) ?: return
    val vote = VoteEntity(chatId, messageId, update.callbackQuery.from.id, update.extractVoteValue())

    if (update.callbackQuery.message.isOld() || meme.senderId == vote.voterId) return

    val votes = voteRepository.getVotes(meme)
        .associate { Pair(it.voterId, it.voteValue) }
        .toMutableMap().also {
          it.merge(vote.voterId, vote.voteValue) { old, new -> if (old == new) null else new }
        }

    val groupedUpVotes = votes.values.groupingBy { it }.eachCount()
    val markup = createMarkup(groupedUpVotes)

    runCatching {
      execute(EditMessageReplyMarkup()
          .setChatId(chatId)
          .setMessageId(messageId)
          .setReplyMarkup(markup))
    }.onFailure {
      log.error("Failed to process vote=$vote", it)
    }

    var privateCaptionPrefix = "мем на модерации. статистика: \n\n"

    if (meme.channelId == null && readyForShipment(votes)) {
      runCatching {
        val originalCaption = update.callbackQuery.message.caption ?: ""
        val captionForChannel = originalCaption.split("Sender: ").firstOrNull()
        execute(SendPhoto()
            .setChatId(channelId)
            .setPhoto(meme.fileId)
            .setCaption(captionForChannel)
            .setReplyMarkup(markup)
        )
      }.onSuccess { message ->
        memeRepository.update(meme, meme.copy(channelId = message.chatId, channelMessageId = message.messageId))
        log.info("Sent meme to channel=$meme")
        privateCaptionPrefix = "мем отправлен на канал. статистика: \n\n"
      }.onFailure {
        log.error("Failed to send meme=$meme to channel", it)
        return
      }
    } else if (meme.channelId != null) {
      runCatching {
        execute(EditMessageReplyMarkup()
            .setChatId(meme.channelId)
            .setMessageId(meme.channelMessageId)
            .setReplyMarkup(markup)
        )
      }.onFailure {
        log.error("Failed to process vote=$vote", it)
        return
      }
    }

    if (votes.containsKey(vote.voterId)) voteRepository.insertOrUpdate(vote) else voteRepository.delete(vote)

    val caption = groupedUpVotes.entries.joinToString(
        prefix = privateCaptionPrefix,
        transform = { (value, sum) -> "${value.emoji}: $sum" })
    updateCaptionInSenderChat(meme, caption)

    log.info("Processed chat vote=$vote")
  }

  private fun processChannelVote(update: Update) {
    val messageId = update.callbackQuery.message.messageId

    val meme = memeRepository.findMemeByChannel(channelId, messageId) ?: return
    val vote = VoteEntity(meme.chatId, meme.messageId, update.callbackQuery.from.id, update.extractVoteValue())

    if (update.callbackQuery.message.isOld() || meme.senderId == vote.voterId) return

    val votes = voteRepository.getVotes(meme)
        .associate { Pair(it.voterId, it.voteValue) }
        .toMutableMap().also {
          it.merge(vote.voterId, vote.voteValue) { old, new -> if (old == new) null else new }
        }

    runCatching {
      val markup = createMarkup(votes.values.groupingBy { it }.eachCount())
      execute(
          EditMessageReplyMarkup()
              .setChatId(channelId)
              .setMessageId(messageId)
              .setReplyMarkup(markup)
      )
      if (meme.chatId != meme.channelId) {
        execute(
            EditMessageReplyMarkup()
                .setChatId(meme.chatId)
                .setMessageId(meme.messageId)
                .setReplyMarkup(markup))
      }
    }.onFailure { throwable ->
      log.error("Failed to process vote=" + vote + ". Exception=" + throwable.message)
      return
    }.onSuccess {
      if (votes.containsKey(vote.voterId)) voteRepository.insertOrUpdate(vote) else voteRepository.delete(vote)

      val caption = votes.values.groupingBy { it }.eachCount().entries.joinToString(
          prefix = "мем отправлен на канал. статистика: \n\n",
          transform = { (value, sum) -> "${value.emoji}: $sum" })
      updateCaptionInSenderChat(meme, caption)

      log.info("Processed channel vote=$vote")
    }
  }

  private fun sendStats(update: Update) {
    fun formatStatsMessage(stats: List<MemeStatsEntry>): String =
        if (stats.isEmpty())
          "You have no statistics yet!"
        else
          """
          Your statistics:
          
          Memes sent: ${stats.size}
        """.trimIndent() +
              stats
                  .flatMap { it.counts.asIterable() }
                  .groupBy({ it.first }, { it.second })
                  .mapValues { it.value.sum() }
                  .toList()
                  .sortedBy { it.first }
                  .joinToString("\n", prefix = "\n\n", transform = { (value, sum) -> "${value.emoji}: $sum" })

    runCatching {
      execute(
          SendMessage()
              .setChatId(update.message.chatId)
              .setText(formatStatsMessage(voteRepository.getStatsByUser(channelId, update.message.from.id)))
      )
    }.onSuccess {
      log.debug("Sent stats to user=${update.message.from.id}")
    }.onFailure { throwable ->
      log.error("Failed to send stats to user=${update.message.from.id}. Exception=", throwable)
    }
  }

  private fun processMeme(update: Update) {
    val message = update.message
    val senderId = message.from.id
    val fileId = message.photo[0].fileId
    val isChatMember = runCatching { execute(GetChatMember().setChatId(chatId).setUserId(senderId)).status }
        .getOrElse { ex ->
          log.error("can't get chat status of user $senderId because of", ex)
          "unknown"
        }.isChatUserStatus()
    val caption = (message.caption?.trim()?.run { this + if (isChatMember) "" else "\n\n" } ?: "") +
        if (isChatMember) ""
        else "Sender: " + (message.from.userName?.let { "@$it" } ?: message.from.firstName ?: message.from.lastName)

    val privateMessageId =
        runCatching {
          execute(DeleteMessage()
              .setChatId(message.chatId)
              .setMessageId(message.messageId))

          execute(SendPhoto()
              .setPhoto(fileId)
              .setChatId(message.chatId)
              .setCaption(message.caption))
        }.getOrThrow().messageId

    fun sendMemeToChat() =
        runCatching {
          execute(
              SendPhoto()
                  .setChatId(chatId)
                  .setPhoto(fileId)
                  .setCaption(caption)
                  .setReplyMarkup(createMarkup(emptyMap())))
        }.onSuccess { sentMessage ->
          memeRepository.save(MemeEntity(chatId, sentMessage.messageId, senderId, fileId, privateMessageId)).also {
            downloadFromFileId(it.fileId).also { file ->
              log.info(file.toString())
              memeMatcher.addMeme(it.fileId, downloadFromFileId(it.fileId))
            }
            log.info("Sent meme=$it to chat")
          }
        }.onFailure {
          log.error("Failed to send meme from message=${message.print()}. Exception=", it)
        }

    fun forwardOriginalMemeToSender(originalFileId: String) =
        runCatching {
          execute(SendMessage()
              .setChatId(update.message.chatId)
              .setReplyToMessageId(update.message.messageId)
              .setText("К сожалению, мем уже был отправлен ранее!"))
          memeRepository.findMeme(originalFileId)?.also {
            if (it.isPublishedOnChannel())
              execute(ForwardMessage()
                  .setChatId(update.message.chatId)
                  .setFromChatId(it.channelId)
                  .setMessageId(it.channelMessageId)
                  .disableNotification()
              )
            else
              execute(ForwardMessage()
                  .setChatId(update.message.chatId)
                  .setFromChatId(it.chatId)
                  .setMessageId(it.messageId)
                  .disableNotification())
          }
        }.onFailure {
          log.error("Failed to forward original meme. Exception=", it)
        }.onSuccess {
          log.info("Successfully forwarded original meme to sender=${message.from.mention()}. $it")
        }

    runCatching {
      downloadFromFileId(fileId).also {
        memeMatcher.tryFindDuplicate(it)?.also { duplicateFileId -> forwardOriginalMemeToSender(duplicateFileId) }
            ?: sendMemeToChat()
      }
    }.onFailure { throwable ->
      log.error("Failed to check duplicates for fileId=$fileId. Message=${message.print()}. Exception=", throwable)
    }
  }

  private fun readyForShipment(votes: MutableMap<Int, VoteValue>): Boolean =
      votes.values.filter { it == UP }.size - votes.values.filter { it == DOWN }.size >= 5

  private fun createMarkup(stats: Map<VoteValue, Int>): InlineKeyboardMarkup {
    fun createVoteInlineKeyboardButton(voteValue: VoteValue, voteCount: Int): InlineKeyboardButton {
      return InlineKeyboardButton().also {
        it.text = if (voteCount == 0) voteValue.emoji else voteValue.emoji + " " + voteCount
        it.callbackData = voteValue.name
      }
    }

    return InlineKeyboardMarkup().setKeyboard(
        listOf(
            listOf(
                createVoteInlineKeyboardButton(UP, stats.getOrDefault(UP, 0)),
                createVoteInlineKeyboardButton(DOWN, stats.getOrDefault(DOWN, 0))
            )
        )
    )
  }

  private fun downloadFromFileId(fileId: String) =
      createTempFile("photo-", "-" + currentThread().id + "-" + currentTimeMillis()).apply { this.deleteOnExit() }.also {
        FileOutputStream(it).use { out ->
          URL(execute(GetFile().setFileId(fileId)).getFileUrl(botToken)).openStream().use { stream -> IOUtils.copy(stream, out) }
        }
      }.also {
        log.info("Successfully downloaded file=$it")
      }
}
